### ISSUE TITLE
Add NodeFieldSelector to EnvVarSource  to add environment generated on the kubelet

### ISSUE THREAD

**Issue by guhan121** [kubernetes/kubernetes#115985]
### What would you like to be added?

# Some Context
But only after the Pod call is completed, we know the node where the Pod is running, and then we can know the availability zone information. 
At this time, it is no longer possible to define environment variables in the spec template.
A tricky way is to add the environment variables we need when rendering environment variables by modifying the kubelet source code.
ex:
```
// Make the environment variables for a pod in the given namespace.
func (kl *Kubelet) makeEnvironmentVariables(pod *v1.Pod, container *v1.Container, podIP string, podIPs []string) ([]kubecontainer.EnvVar, error) {
	...
	if v, ok := kl.nodeLabels["topology.kubernetes.io/zone"]; ok && len(v) > 0 {
		result = append(result, kubecontainer.EnvVar{Name: "MY_K8S_ZONE", Value: v})
	}

	return result, nil
}

```
if we have a `NodeFliedSelector`, We can easily obtain node-related information into pod environment variables.
```
apiVersion: v1
kind: Pod
spec:
  containers:
    - name: test-container
      image: registry.k8s.io/busybox
      command: [ "sh", "-c"]
      args:
      - while true; do sleep 10; done;
      env:
        - name: MY_K8S_ZONE
          valueFrom:
            nodeFeildRef:
              fieldPath: metadata.labels['topology.kubernetes.io/zone']
   
```

### Why is this needed?

As said before, we need to be able to confirm some environment variables into the container after scheduling.

**Comment by tzneal** [comment]
I think this is a duplicate of https://github.com/kubernetes/kubernetes/issues/40610
